src/kernel/bitmap.zig

```diff
@@ -0,0 +1,228 @@
+
+        num_bitmaps: u32,
+        num_entries: u32,
+        bitmaps: []BitmapType,
+        num_free_entries: u32,
+
+        pub fn init(num_entries: u32, allocator: *std.mem.Allocator) !Self {
+            const num = num_entries / ENTRIES_PER_BITMAP + @boolToInt(num_entries % ENTRIES_PER_BITMAP != 0);
+            const self = Self{
+                .num_bitmaps = num,
+                .num_entries = num_entries,
+                .bitmaps = try allocator.alloc(BitmapType, num),
+                .num_free_entries = num_entries,
+            };
+            for (self.bitmaps) |*bmp| {
+                bmp.* = 0;
+            }
+            return self;
+        }
+
+        ///
+        /// Set an entry within a bitmap as occupied.
+        ///
+        /// Arguments:
+        ///     OUT bmp: *Bitmap - The bitmap to modify.
+        ///     IN idx: BitmapIndex - The index within the bitmap to set.
+        ///
+        pub fn setEntry(self: *Self, idx: u32) BitmapError!void {
+            if (idx >= self.num_entries) return BitmapError.OutOfBounds;
+            if (!try self.isSet(idx)) {
+                const bit = @as(BitmapType, 1) << @intCast(IndexType, idx % ENTRIES_PER_BITMAP);
+                self.bitmaps[idx / ENTRIES_PER_BITMAP] |= bit;
+                self.num_free_entries -= 1;
+            }
+        }
+
+        pub fn clearEntry(self: *Self, idx: u32) BitmapError!void {
+            if (idx >= self.num_entries) return BitmapError.OutOfBounds;
+            if (try self.isSet(idx)) {
+                const bit = @as(u32, 1) << @intCast(u5, idx % ENTRIES_PER_BITMAP);
+                self.bitmaps[idx / ENTRIES_PER_BITMAP] &= ~bit;
+                self.num_free_entries += 1;
+            }
+        }
+
+        ///
+        /// Set the first free entry within the bitmaps as occupied.
+        ///
+        /// Return: The index within all bitmaps that was set.
+        ///     0 .. ENTRIES_PER_BITMAP - 1 if in the first bitmap, ENTRIES_PER_BITMAP .. ENTRIES_PER_BITMAP * 2 - 1 if in the second etc.
+        ///
+        pub fn setFirstFree(self: *Self) ?u32 {
+            if (self.num_free_entries == 0) return null;
+            for (self.bitmaps) |*bmp, i| {
+                if (bmp.* == BITMAP_FULL)
+                    continue;
+                const bit = @truncate(IndexType, @ctz(BitmapType, ~bmp.*));
+                const idx = bit + @intCast(u32, i) * ENTRIES_PER_BITMAP;
+                // Failing here means that the index is outside of the bitmap, so there are no free entries
+                self.setEntry(idx) catch return null;
+                return idx;
+            }
+            return null;
+        }
+
+        pub fn isSet(self: *Self, idx: u32) BitmapError!bool {
+            if (idx >= self.num_entries) return BitmapError.OutOfBounds;
+            return (self.bitmaps[idx / ENTRIES_PER_BITMAP] & @as(u32, 1) << @intCast(u5, idx % ENTRIES_PER_BITMAP)) != 0;
```

Comment:
`@as(BitmapType, 1) << @intCast(IndexType, idx % ENTRIES_PER_BITMAP)`. So maybe this could be a function as it is used in 3 places